store/src/java/com/zimbra/cs/lmtpserver/ExternalEmailWarning.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * Get configured warning to be used on text/html mime message
+     * parts.
+     * 
+     * @return string with warning note.
+     */
+    public String getTextHtmlWarning() {
+        return textHtmlWarning;
+    }
+
+    /**
+     * Get configured warning to be used on text/plain mime message
+     * parts.
+     * 
+     * @return string with warning note.
+     */
+    public String getTextPlainWarning() {
+        return textPlainWarning;
+    }
+
+    private static final String EMAIL_REGEX = "^(.+)@(\\S+)$";
+    private static final Pattern EMAIL_PATTERN = Pattern.compile(EMAIL_REGEX);
+
+    /**
+     * Finds whether a originator is external to a receiver's organization based on
+     * the string representation of their addresses.
+     * 
+     * If their domains are equal or at least one is sub-domain to the other, then
+     * the addresses are considered as intra-organization; otherwise, the addresses
+     * are considered as extra-organization.
+     * 
+     * This method works under an optimistic behavior: shall any of the arguments
+     * represent an invalid email address, or shall anything goes wrong in any other
+     * aspect with the analysis, it will assume that the originator is not external
+     * and thus return false.
+     * 
+     * Please note that no actual directory (LDAP) validations are done here, just
+     * analysis of two independent as-is strings.
+     * 
+     * @param receiverAddress
+     *            receiver's email address in RFC5322 format: name@domain.tld
+     * 
+     * @param originatorAddress
+     *            originator's email address in RFC5322 format: name@domain.tld
+     * 
+     * @return true if originatorAddress is external to receiverAddress domain
+     */
+    public boolean isExternal(String receiverAddress, String originatorAddress) {
+        // finding receiverAddress validity
+        final Matcher receiverAddressMatcher = EMAIL_PATTERN.matcher(receiverAddress);
+        if (receiverAddressMatcher.matches()) {
+            final String receiverDomain = receiverAddressMatcher.group(2);
+            // finding originatorAddress validity
+            final Matcher originatorAddressMatcher = EMAIL_PATTERN.matcher(originatorAddress);
+            if (originatorAddressMatcher.matches()) {
+                final String originatorDomain = originatorAddressMatcher.group(2);
+                // checking domain names are correctly identified for added safety
+                if (receiverDomain != null && originatorDomain != null) {
+                    // splitting domain names into segments
+                    final List<String> receiverDomainSegments = Arrays.asList(receiverDomain.split("\\."));
+                    final List<String> originatorDomainSegments = Arrays.asList(originatorDomain.split("\\."));
+                    // reversing to analyze from highest-hierarchy segment (TLD)
+                    Collections.reverse(receiverDomainSegments);
```

Comment:
Is it really need to reverse the collections. Can't do like this?
int size = Math.min(receiverDomainSegments.size(), originatorDomainSegments.size());
for (int i = size - 1; i <= 0; i --) {
.......
.......
}

Reply:
Agreed! no need to reverse.

Done.